source/common/formatter/substitution_formatter.cc

```diff
@@ -1918,6 +1926,48 @@ UpstreamHostMetadataFormatter::UpstreamHostMetadataFormatter(const std::string&
                           return host->metadata().get();
                         }) {}
 
+// Parses the proxy protocol TLVs stored in the FilterState with the key "envoy.network.proxy_protocol_options" 
+// and returns a function suitable for accessing the specified metadata from an StreamInfo::StreamInfo. 
+absl::optional<std::string> ProxyProtocolTlvsFormatter::format(const Http::RequestHeaderMap&,
+                                                        const Http::ResponseHeaderMap&,
+                                                        const Http::ResponseTrailerMap&,
+                                                        const StreamInfo::StreamInfo& stream_info,
+                                                        absl::string_view) const {
+  const Envoy::StreamInfo::FilterState& filter_state = stream_info.filterState();
+  auto typed_state = filter_state.getDataReadOnly<Network::ProxyProtocolFilterState>(Network::ProxyProtocolFilterState::key());
+
+    // Value exists but isn't string accessible is a contract violation; throw an error.
+    if (typed_state == nullptr) {
+      ENVOY_LOG_MISC(debug, "Invalid header: PROXY_PROTOCOL_TLVS.");
+      return std::string();
+    }
+
+  // Format the vector of proxy protocol TLVs as follows: <type integer>|<base64 encoded value>[,...]
+    std::ostringstream oss;
+  if (!typed_state->value().tlv_vector_.empty()) {
+      ENVOY_LOG_MISC(debug, "parsing proxy protocol TLVs: length of the TLV vector is {}", typed_state->value().tlv_vector_.size());
+
+      // Add the first element with no delimiter
+      auto first_tlv = typed_state->value().tlv_vector_.cbegin();
+
+      // Base64 encode the TLV value
+      auto base64_encoded_tlv_value = Base64::encode(first_tlv->value.data(), first_tlv->value.length());
+      oss << fmt::format("{}|{}", first_tlv->type, base64_encoded_tlv_value);
+      while (typed_state->value().tlv_vector_.cend() != ++first_tlv) {
+        oss << fmt::format(", {}|{}", first_tlv->type, Base64::encode(first_tlv->value.data(), first_tlv->value.length()));
+      }
+  }
```

Comment:
[q] From my current manual testing with an adapted version of the proxy protocol server/client POC in serverless-proxy repo, the following headers are received by envoy: 

```
[2023-01-11 21:39:41.038][19911][debug][router] [source/common/router/router.cc:678] [C38][S16547806752358855165] router decoding headers:
':authority', 'pong.foo.bar'
':path', '/'
':method', 'POST'
':scheme', 'https'
'content-length', '4'
'x-forwarded-proto', 'https'
'x-request-id', '657e4297-c041-4ce5-9461-d368edac551f'
'x-envoy-expected-rq-timeout-ms', '15000'
'proxy-protocol-tlv', '5|AAAAAAAAAAAAAAAAAAAAAQ=='
```
Does the encoded proxy protocol TLV value look correct? I'm not entirely sure how the value should look after it's encoded, and the encoders/decoders that I've used to try to figure out how it should look don't agree with the `AAAAAAAAAAAAAAAAAAAAAQ==` value... I'm not sure if the issue is the encoding method or how the TLV vector is getting accessed from the FilterState.

Reply:
This [decoder](https://base64.guru/converter/decode/file) makes me think this _might_ be the expected value? 

<img width="1353" alt="image" src="https://user-images.githubusercontent.com/36177962/211927469-5f704922-ef86-425c-9594-f3747917e759.png">

Have never had to read base64 before 😅

Reply:
I think this is the right base64 encoded value, so I'm going to close this comment :)